stub/src/main/java/io/grpc/kotlin/ServerCalls.kt

```diff
@@ -238,34 +236,30 @@ object ServerCalls {
     }
 
     val rpcJob = CoroutineScope(context).launch {
-      val mutex = Mutex()
-      val headersSent = AtomicBoolean(false) // enforces only sending headers once
+      var headersSent = false
       val failure = runCatching {
         implementation(requests).collect {
           // once we have a response message, check if we've sent headers yet - if not, do so
-          if (headersSent.compareAndSet(false, true)) {
-            mutex.withLock {
-              call.sendHeaders(GrpcMetadata())
-            }
+          if (!headersSent) {
+            call.sendHeaders(GrpcMetadata())
+            headersSent = true
           }
           readiness.suspendUntilReady()
-          mutex.withLock { call.sendMessage(it) }
+          call.sendMessage(it)
         }
       }.exceptionOrNull()
       // check headers again once we're done collecting the response flow - if we received
       // no elements or threw an exception, then we wouldn't have sent them
-      if (failure == null && headersSent.compareAndSet(false, true)) {
-        mutex.withLock {
-          call.sendHeaders(GrpcMetadata())
-        }
+      if (failure == null && !headersSent) {
```

Comment:
I didn't change this logic. But it's now a bit more clear.

If the RPC didn't throw and sent nothing, then send blank headers.

If the RPC throws, then it won't sendHeaders

The comment above it indicates a different expectation